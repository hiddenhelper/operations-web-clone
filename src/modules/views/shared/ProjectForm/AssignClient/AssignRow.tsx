import React, { memo, useMemo, useCallback, useEffect } from 'react';
import { Link } from 'react-router-dom';

import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import FormControlLabel from '@material-ui/core/FormControlLabel';

import AssignEntity from '../../AssignEntity';
import DeleteEntityButton from '../../DeleteEntityButton';
import Checkbox from '../../FormHandler/Checkbox';
import ControlledError from '../../FormHandler/ControlledError';

import { IAssignEntityProps } from '../../AssignEntity/AssignEntity';
import { ResourceModel } from '../../../../models';
import { getConditionalDefaultValue } from '../../../../../utils/generalUtils';
import { getDrawerButton } from '../../../../../utils/clientUtils';
import { useStyles } from '../styles';

export interface IAssignClientRowProps {
  index: number;
  isNew: boolean;
  duplicity: string;
  error: string;
  assignEntityProps: IAssignEntityProps;
  entityResponse: { id: string; name: string; status?: number };
  hideBottom?: boolean;
  showTaxExempt?: boolean;
  onDelete: (id: number, tempId: string) => void;
  onUpdate: (index: number, value: boolean) => void;
}

const AssignClientRow = ({
  isNew,
  error,
  duplicity,
  assignEntityProps,
  index,
  entityResponse,
  hideBottom = false,
  showTaxExempt,
  onDelete,
  onUpdate,
}: IAssignClientRowProps) => {
  const classes = useStyles();
  const statusListToShowResponse = useMemo(
    () => [
      ResourceModel.CompanyStatus.DRAFT,
      ResourceModel.CompanyStatus.PENDING_APPROVAL,
      ResourceModel.CompanyStatus.ARCHIVED,
      ResourceModel.CompanyStatus.ACTIVE,
    ],
    []
  );
  const showEntityResponse = useMemo(() => entityResponse && statusListToShowResponse.includes(entityResponse.status), [
    entityResponse,
    statusListToShowResponse,
  ]);
  const onDeleteHandler = useCallback(() => onDelete(index, assignEntityProps.tempId), [index, onDelete]); // eslint-disable-line
  const duplicityMsg = useMemo(() => (duplicity ? duplicity : ''), [duplicity]);
  const errorMsg = useMemo(() => (error ? error : duplicityMsg), [duplicityMsg, error]);

  const handleTaxExemptChange = useCallback(
    event => {
      onUpdate(index, event.target.checked);
    },
    [onUpdate, index]
  );

  useEffect(() => {
    if (isNew && entityResponse) assignEntityProps.onSelect(index, entityResponse, assignEntityProps.tempId);
  }, [index, entityResponse, isNew]); // eslint-disable-line
  return (
    <Grid
      key={`${assignEntityProps.tempId}-${index}`}
      container={true}
      className={`${classes.formControlWrapper} ${getConditionalDefaultValue(hideBottom, classes.rowsWrapperWithoutButton, classes.rowsWrapper)}`}
    >
      <Grid item={true} xs={11} className={classes.inputWrapper}>
        <ControlledError show={!!errorMsg} error={errorMsg}>
          <AssignEntity {...assignEntityProps} />
          {!duplicity && showEntityResponse && (
            <Typography data-testid="entity-row" className={classes.entityResponse}>
              "{entityResponse.name}" is <span className={classes.accent}>{ResourceModel.statusMap[entityResponse.status]}</span>.
              {entityResponse.status !== ResourceModel.CompanyStatus.ACTIVE && (
                <>
                  {' '}
                  You can add more information and approve it in{' '}
                  <Link className={`${classes.accent} ${classes.editAccentColor}`} to={getDrawerButton(entityResponse.status, entityResponse.id).linkTo}>
                    Edit Client
                  </Link>
                </>
              )}
            </Typography>
          )}
        </ControlledError>
        {showTaxExempt && (
          <FormControlLabel
            className={classes.taxExemptCheckbox}
            label="Tax exempt"
            disabled={!assignEntityProps?.assignValue?.name}
            control={
              <Checkbox
                name="isTaxExempt"
                value="isTaxExempt"
                isChecked={assignEntityProps?.assignValue?.isTaxExempt}
                onChange={handleTaxExemptChange}
                inputProps={{ 'data-testid': 'client-isTaxExempt' }}
                disabled={!assignEntityProps?.assignValue?.name}
              />
            }
          />
        )}
      </Grid>
      <Grid item={true}>
        <DeleteEntityButton
          styleClass={`${assignEntityProps && assignEntityProps.assignValue.name.length ? classes.showDeleteButton : classes.hideDeleteButton}`}
          onClick={onDeleteHandler}
        />
      </Grid>
    </Grid>
  );
};

export default memo(AssignClientRow);
