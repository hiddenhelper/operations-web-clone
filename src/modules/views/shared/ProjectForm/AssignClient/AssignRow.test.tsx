import React from 'react';
import { render, act, fireEvent, RenderResult } from '@testing-library/react';
import { MemoryRouter } from 'react-router';

import AssignRow, { IAssignClientRowProps } from './AssignRow';

jest.mock('@material-ui/lab/Autocomplete', () => {
  return (() => ({ value, onChange }) => {
    return (
      <div>
        <input data-testid="assign-entity-item" value={value} onChange={onChange} />
      </div>
    );
  })();
});

describe('AssignRow', () => {
  let wrapper: RenderResult;
  let props: IAssignClientRowProps;

  beforeEach(() => {
    props = {
      index: 0,
      isNew: false,
      duplicity: null,
      error: null,
      assignEntityProps: {
        index: 0,
        tempId: 'temp-id-0',
        result: [],
        params: { isDeveloper: false },
        optionLabel: 'name',
        assignValue: { name: 'value' },
        isLoading: false,
        existRelation: false,
        showCreateNew: true,
        placeholder: '',
        renderOption: jest.fn(),
        search: jest.fn(),
        onCreate: jest.fn(),
        onReset: jest.fn(),
        onSelect: jest.fn(),
      },
      entityResponse: {} as any,
      onDelete: jest.fn(),
    };
  });

  it('should render', () => {
    wrapper = render(<AssignRow {...props} />);
    expect(wrapper.container).toMatchSnapshot();
  });

  it('should delete row', () => {
    wrapper = render(<AssignRow {...props} />);
    const deleteBtn = wrapper.getByTestId('delete-entity-button');
    act(() => {
      fireEvent.click(deleteBtn);
    });

    expect(props.onDelete).toHaveBeenCalledWith(0, 'temp-id-0');
  });

  it('should show row label', () => {
    props.entityResponse = { id: 'id', name: 'row name', status: 0 };
    wrapper = render(
      <MemoryRouter>
        <AssignRow {...props} />
      </MemoryRouter>
    );
    const deleteBtn = wrapper.getByTestId('delete-entity-button');
    act(() => {
      fireEvent.click(deleteBtn);
    });

    expect(props.onDelete).toHaveBeenCalledWith(0, 'temp-id-0');
  });

  it('should show duplicity message', () => {
    props.duplicity = 'is duplicated';
    wrapper = render(
      <MemoryRouter>
        <AssignRow {...props} />
      </MemoryRouter>
    );
    expect(wrapper.container).toMatchSnapshot();
  });

  it('should show error message', () => {
    props.error = 'error message';
    props.assignEntityProps.assignValue = { name: '' };
    wrapper = render(
      <MemoryRouter>
        <AssignRow {...props} />
      </MemoryRouter>
    );
    expect(wrapper.container).toMatchSnapshot();
  });

  it('should set entity as selected', () => {
    props.isNew = true;
    props.entityResponse = { id: 'id', name: 'row name', status: 0 };
    wrapper = render(
      <MemoryRouter>
        <AssignRow {...props} />
      </MemoryRouter>
    );
    expect(props.assignEntityProps.onSelect).toHaveBeenCalledWith(0, props.entityResponse, 'temp-id-0');
  });
});
