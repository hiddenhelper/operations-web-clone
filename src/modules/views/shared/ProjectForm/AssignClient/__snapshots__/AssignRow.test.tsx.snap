// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssignRow should render 1`] = `
<div>
  <div
    class="MuiGrid-root makeStyles-formControlWrapper-52 makeStyles-rowsWrapper-12 MuiGrid-container"
  >
    <div
      class="MuiGrid-root makeStyles-inputWrapper-53 MuiGrid-item MuiGrid-grid-xs-11"
    >
      <div
        class="MuiFormControl-root makeStyles-formControl-81 makeStyles-autoCompleteWrapper-73 "
      >
        <h1
          class="MuiTypography-root MuiTypography-caption MuiTypography-colorSecondary MuiTypography-alignLeft"
        >
          Client Name
           
        </h1>
        <div>
          <div>
            <input
              data-testid="assign-entity-item"
              value="[object Object]"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item"
    >
      <span
        class=""
        data-testid="controlled-button"
        tabindex="-1"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root makeStyles-deleteButton-83 makeStyles-showDeleteButton-51"
          data-testid="delete-entity-button"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg>
              delete.svg
            </svg>
          </span>
        </button>
      </span>
    </div>
  </div>
</div>
`;

exports[`AssignRow should show duplicity message 1`] = `
<div>
  <div
    class="MuiGrid-root makeStyles-formControlWrapper-301 makeStyles-rowsWrapper-261 MuiGrid-container"
  >
    <div
      class="MuiGrid-root makeStyles-inputWrapper-302 MuiGrid-item MuiGrid-grid-xs-11"
    >
      <div
        class="MuiFormControl-root makeStyles-formControl-330 makeStyles-autoCompleteWrapper-322 "
      >
        <h1
          class="MuiTypography-root MuiTypography-caption MuiTypography-colorSecondary MuiTypography-alignLeft"
        >
          Client Name
           
        </h1>
        <div>
          <div>
            <input
              data-testid="assign-entity-item"
              value="[object Object]"
            />
          </div>
        </div>
      </div>
      <h1
        class="MuiTypography-root  makeStyles-errorMessage-321 errorMessage MuiTypography-caption MuiTypography-colorSecondary MuiTypography-alignLeft"
      >
        is duplicated
      </h1>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item"
    >
      <span
        class=""
        data-testid="controlled-button"
        tabindex="-1"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root makeStyles-deleteButton-332 makeStyles-showDeleteButton-300"
          data-testid="delete-entity-button"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg>
              delete.svg
            </svg>
          </span>
        </button>
      </span>
    </div>
  </div>
</div>
`;

exports[`AssignRow should show error message 1`] = `
<div>
  <div
    class="MuiGrid-root makeStyles-formControlWrapper-384 makeStyles-rowsWrapper-344 MuiGrid-container"
  >
    <div
      class="MuiGrid-root makeStyles-inputWrapper-385 MuiGrid-item MuiGrid-grid-xs-11"
    >
      <div
        class="MuiFormControl-root makeStyles-formControl-413 makeStyles-autoCompleteWrapper-405 "
      >
        <h1
          class="MuiTypography-root MuiTypography-caption MuiTypography-colorSecondary MuiTypography-alignLeft"
        >
          Client Name
           
        </h1>
        <div>
          <div>
            <input
              data-testid="assign-entity-item"
              value="[object Object]"
            />
          </div>
        </div>
      </div>
      <h1
        class="MuiTypography-root  makeStyles-errorMessage-404 errorMessage MuiTypography-caption MuiTypography-colorSecondary MuiTypography-alignLeft"
      >
        error message
      </h1>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item"
    >
      <span
        class=""
        data-testid="controlled-button"
        tabindex="-1"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root makeStyles-deleteButton-415 makeStyles-hideDeleteButton-382"
          data-testid="delete-entity-button"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg>
              delete.svg
            </svg>
          </span>
        </button>
      </span>
    </div>
  </div>
</div>
`;
